pkg/sql/authorization.go

```diff
@@ -451,3 +463,29 @@ const ConnAuditingClusterSettingName = "server.auth_log.sql_connections.enabled"
 // counts in SetClusterSetting() and importing pgwire here would
 // create a circular dependency.
 const AuthAuditingClusterSettingName = "server.auth_log.sql_sessions.enabled"
+
+func (p *planner) canCreateOnSchema(
+	ctx context.Context, dbID descpb.ID, schemaID descpb.ID, user string,
+) error {
+	resolvedSchema, err := resolver.ResolveSchemaDescByID(ctx, p.Txn(), p.ExecCfg().Codec, schemaID)
+	if err != nil {
+		return err
+	}
+
+	switch resolvedSchema.Kind {
+	case sqlbase.SchemaPublic:
+		// Anyone can CREATE on a public schema.
+		return nil
+	case sqlbase.SchemaTemporary, sqlbase.SchemaVirtual:
+		scName, err := resolver.ResolveSchemaNameByID(ctx, p.Txn(), p.ExecCfg().Codec, dbID, schemaID)
+		if err != nil {
+			return err
+		}
+		return pgerror.Newf(pgcode.InsufficientPrivilege,
+			"cannot CREATE on schema %s", scName)
```

Comment:
should this be "within" rather than on?

Reply:
I think I'd actually prefer leaving this as on since we do `GRANT x ON SCHEMA` and our regular check privilege error will say ON for schema as well.